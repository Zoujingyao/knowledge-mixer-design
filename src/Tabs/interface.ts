import React from 'react';

export type TabItemType = {
  /** 对应 activeKey */
  key: string;
  /** 选项卡头显示文字 */
  label: React.ReactNode;
  /** 选项卡面板显示内容 */
  children: React.ReactNode;
  /** 是否禁用该面板 */
  disabled?: boolean;
};

export type TabsProps = {
  /** 配置选项卡内容 */
  items?: TabItemType[];
  /** 当前激活 tab 面板的 key, 默认激活第一项 */
  activeKey?: string;
  /** 切换面板的回调 */
  onChange?: (activeKey: string) => void;
  /** 标签居中展示 */
  centered?: boolean;
  /** tab bar 上额外的元素 */
  tabBarExtraContent?: React.ReactNode;
  /** 面板样式类名 */
  panelClassName?: string;
};

export type TabNavProps = Omit<TabsProps, 'items'> & {
  tabs?: TabItemType[];
};

export type TabContext = {
  selectedKey: string;
  setSelectedKey: (selectedKey: string) => void;
  tabs: TabItemType[];
  setTabs: (tabs: TabItemType[]) => void;
};
