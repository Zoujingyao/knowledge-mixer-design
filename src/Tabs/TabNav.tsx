import React, { useContext } from 'react';
import classNames from 'classnames';
import type { TabItemType, TabNavProps } from './interface';
import { TabsContext } from './index';
import { getPrefixCls } from '@/utils';
import './index.less';

const TabNav: React.FC<TabNavProps> = (props) => {
  const { selectedKey, tabs, setSelectedKey } = useContext(TabsContext);

  const onTabChange = (tab: TabItemType) => {
    if (tab?.disabled) return;
    setSelectedKey(tab.key);
  };

  const tabNavPrefixCls = `${getPrefixCls('tab')}-nav`;
  const navWrapPrefixCls = classNames(tabNavPrefixCls, {
    [`${tabNavPrefixCls}-centered`]: props?.centered,
  });
  const tabNavListPrefixCls = `${tabNavPrefixCls}-list`;
  const tabNavItemPrefixCls = `${tabNavListPrefixCls}-item`;

  return (
    <div className={navWrapPrefixCls}>
      <div className={tabNavListPrefixCls}>
        {tabs?.map((tab) => {
          const itemClassName = classNames(tabNavItemPrefixCls, {
            [`${tabNavItemPrefixCls}-selected`]: selectedKey === tab.key,
            [`${tabNavItemPrefixCls}-disabled`]: tab?.disabled || false,
          });
          return (
            <div key={tab.key} onClick={() => onTabChange(tab)} className={itemClassName}>
              {tab.label}
            </div>
          );
        })}
      </div>
      {props?.tabBarExtraContent}
    </div>
  );
};

export default TabNav;
