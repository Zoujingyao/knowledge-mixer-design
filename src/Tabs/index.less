.km-tab {
  &-nav {
    display: flex;
    justify-content: space-between;
    width: 100%;

    &-centered {
      justify-content: center;
    }

    &-list {
      display: inline-block;
      box-sizing: content-box;
      margin-bottom: 16px;
      padding: 2px;
      font-weight: 500;
      letter-spacing: 1px;
      background: rgba(22, 22, 64, 0.04);
      border: 1px solid rgba(22, 22, 64, 0.08);
      border-radius: 6px;

      &-item {
        display: inline-block;
        height: 32px;
        color: @primary;
        font-size: 14px;
        line-height: 32px;
        border: none;
        border-radius: 4px;
        cursor: pointer;
        transition: color 0.2s, background 0.2s, border-color 0.2s, box-shadow 0.2s;
        padding-inline: 15px;
        padding-block: 0;

        &-selected {
          background: #fff;
        }

        &-disabled {
          color: @gray500;
          cursor: not-allowed;
        }
      }
    }
  }

  &-panel {
    margin-top: 20px;
  }
}
