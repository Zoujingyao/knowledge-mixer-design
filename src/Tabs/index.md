---
title: Tabs 标签页
nav:
  title: Components
  path: /components
group:
  title: 数据展示
  path: /components/data-display
---

# Tabs 标签页

## 基本使用
<code src="./demo/index.tsx"></code>
