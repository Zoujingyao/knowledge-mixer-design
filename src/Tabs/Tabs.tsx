import React, { useContext } from 'react';
import { TabsProps } from './interface';
import TabNav from './TabNav';
import { TabsContext } from './index';
import styles from './index.less';
import classNames from 'classnames';

const Tabs: React.FC<TabsProps> = (props) => {
  const { selectedKey, tabs } = useContext(TabsContext);

  const activeIdx: number | undefined = React.useMemo(() => {
    if (!tabs.length) return undefined;
    if (!selectedKey) return 0;
    return tabs?.findIndex((tab) => tab.key === selectedKey);
  }, [selectedKey, tabs]);

  return (
    <div className={styles.kmTab}>
      <TabNav {...props} />
      {activeIdx !== undefined && (
        <div className={classNames(styles.kmTabPanel, props?.panelClassName)}>{tabs?.[activeIdx]?.children}</div>
      )}
    </div>
  );
};

export default Tabs;
