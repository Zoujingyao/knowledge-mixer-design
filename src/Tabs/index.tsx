import React, { useEffect, useState } from 'react';
import { TabContext, TabItemType, TabsProps } from './interface';
import Tabs from './Tabs';

export const TabsContext = React.createContext<TabContext>({
  selectedKey: '',
  setSelectedKey: () => {
    return;
  },
  tabs: [],
  setTabs: () => {
    return;
  },
});

function TabsWrapper(props: React.PropsWithChildren<TabsProps>) {
  const { activeKey, items } = props;

  const [tabs, setTabs] = useState<TabItemType[]>([]);
  const [selectedKey, setSelectedKey] = useState<string>(activeKey || '');

  useEffect(() => {
    setTabs((items || []).filter((item) => item && typeof item === 'object' && 'key' in item));
  }, [items]);

  useEffect(() => {
    if (!selectedKey) setSelectedKey(tabs?.[0]?.key);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [tabs]);

  return (
    <TabsContext.Provider
      value={{
        selectedKey,
        setSelectedKey,
        tabs,
        setTabs,
      }}
    >
      <Tabs {...props} />
    </TabsContext.Provider>
  );
}

export default TabsWrapper;
