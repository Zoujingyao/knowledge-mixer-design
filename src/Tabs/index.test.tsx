import React from 'react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import Tabs from './index';

describe('<Tabs />', () => {
  it('render Tabs with dumi', () => {
    try {
      const ele = renderer
        .create(
          <Tabs
            items={new Array(3).fill(null).map((_, i) => {
              const id = String(i + 1);
              return {
                label: `Tab ${id}`,
                key: id,
                children: `Content of Tab Pane ${id}`,
              };
            })}
          />,
        )
        .toJSON();
      expect(ele).toMatchSnapshot();
    } catch (e) {
      console.log(e);
    }
  });
});
