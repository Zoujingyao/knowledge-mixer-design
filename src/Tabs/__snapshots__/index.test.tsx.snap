// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> render Tabs with dumi 1`] = `
<div
  className="kmTab"
>
  <div
    className="km-tab-nav"
  >
    <div
      className="km-tab-nav-list"
    />
  </div>
</div>
`;
